File: backend/routes/answer.py
from datetime import datetime

from flask import Blueprint, request, current_app
from flask_jwt_extended import jwt_required, get_jwt_identity

from services import answer_service
from services.user_profile_service import UserProfileService
from utils.api_response import success, fail
from utils.validators import parse_page_limit, validate_enum

answer_bp = Blueprint('answer', __name__)


@answer_bp.route('/ask', methods=['POST'])
@jwt_required()
def ask_answer():
    data = request.get_json(silent=True)
    question = (data or {}).get('question')

    if not isinstance(question, str) or not question.strip():
        return fail("question 必须为非空字符串", code=400)
    if len(question.strip()) > 200:
        return fail("question 长度不能超过200", code=400)

    user_id = get_jwt_identity()
    payload = answer_service.ask_question(user_id=user_id, question=question)
    try:
        UserProfileService.update_profile_by_behavior(
            user_id=user_id, event_type="answer_created", event_time=datetime.utcnow()
        )
    except Exception:
        current_app.logger.warning("profile rule fields update failed", exc_info=True)
    return success(data=payload, message="success", code=200)


@answer_bp.route('/history', methods=['GET'])
@jwt_required()
def answer_history():
    try:
        page, limit = parse_page_limit(request.args)
    except ValueError as err:
        return fail(str(err), code=400)

    user_id = get_jwt_identity()
    payload = answer_service.list_history(user_id=user_id, page=page, limit=limit)
    return success(data=payload, message="success", code=200)


@answer_bp.route('/favorite', methods=['POST'])
@jwt_required()
def toggle_favorite():
    data = request.get_json(silent=True)
    if not isinstance(data, dict):
        return fail("请求体必须是有效的JSON", code=400)

    answer_id = data.get('answerId')
    action = data.get('action')

    if not isinstance(answer_id, str) or not answer_id.strip():
        return fail("answerId 字段不能为空", code=400)

    try:
        validate_enum(action, ("favorite", "unfavorite"), "action")
    except ValueError as err:
        return fail(str(err), code=400)

    user_id = get_jwt_identity()
    try:
        payload = answer_service.toggle_favorite(
            user_id=user_id,
            answer_id=answer_id.strip(),
            action=action,
        )
    except LookupError as err:
        return fail(str(err), code=404)

    return success(data=payload, message="success", code=200)

File: frontend/src/api/answer.ts
// src/api/answer.ts
import request from '@/utils/request'

// ========== 类型定义 ==========
export interface AnswerHistoryItem {
  id: string
  question: string
  answerText: string
  createdAt: string
  isFavorited?: boolean
}

/** 传统分页响应结构 */
export interface AnswerHistoryPage {
  total: number
  page: number
  limit: number
  list: AnswerHistoryItem[]
}

/** 提交提问响应 */
export interface AskResponse {
  id: string
  question: string
  answerText: string
  createdAt: string
}

/** 收藏/取消收藏响应 */
export interface FavoriteResponse {
  answerId: string
  isFavorited: boolean
}

// ========== API 函数 ==========

/**
 * 提交提问并抽取答案
 * POST /v1/answer/ask
 * @param question 问题内容（1-200字符）
 */
export const askQuestion = (question: string): Promise<AskResponse> => {
  return request.post('/v1/answer/ask', { question })
}

/**
 * 获取历史提问记录（分页）
 * GET /v1/answer/history
 * @param page 页码，默认 1
 * @param limit 每页数量，默认 10，最大 50
 */
export const getAnswerHistory = (page = 1, limit = 10): Promise<AnswerHistoryPage> => {
  return request.get('/v1/answer/history', { params: { page, limit } })
}

/**
 * 收藏或取消收藏答案
 * POST /v1/answer/favorite
 * @param answerId 答案记录ID
 * @param action favorite | unfavorite
 */
export const favoriteAnswer = (
  answerId: string,
  action: 'favorite' | 'unfavorite',
): Promise<FavoriteResponse> => {
  return request.post('/v1/answer/favorite', { answerId, action })
}

/**
 * 获取收藏的答案列表
 * GET /v1/user/history/favorites （属于用户模块，但为了统一放在 answer.ts 或 user.ts 均可）
 * @param page 页码，默认 1
 * @param limit 每页数量，默认 10
 */
export const getFavoriteAnswers = (page = 1, limit = 10): Promise<AnswerHistoryPage> => {
  return request.get('/v1/user/history/favorites', { params: { page, limit } })
}
